### PR TITLE
Basic game done

### PR DESCRIPTION
**Pull request by mark-a**
Happy review!

### CODE REVIEW

File: lib/game/board.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    @fields = Matrix.build(num_rows, num_cols) {EMPTY_FIELD}
+    @win_condition_length = winning_length
+  end
+
+  def insert_char_at(character, column_number)
+    raise 'Invalid column number' if column_number < 1 || column_number > @fields.column_count
+    x = column_number - 1
+    y = find_drop_point x
+    raise 'Column full' if y < 0
+    @fields.send(:[]=, y, x, character)
```
Comment (splattael) [inline lib/game/board.rb]: As `Matrix` is immutable, you have to use `send` here.
How about creating a new `Matrix`?

File: lib/game/board.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    (fields.row_count - 1).downto 0 do |x|
+      0.upto(fields.column_count - 1) do |y|
+        return true if winning_line_at? x, y
+      end
+    end
+
+    false
+  end
+
+  def winning_line_at?(row, col)
```
Comment (splattael) [inline lib/game/board.rb]: :+1:

File: lib/game/connect_four_game.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
+require 'state_machines'
+
+class ConnectFourGame
+
+  SYMBOL_PLAYER_1 = 'x'.freeze
+  SYMBOL_PLAYER_2 = 'o'.freeze
+
+  attr_accessor :board
+  state_machine :state, initial: :new do
```
Comment (splattael) [inline lib/game/connect_four_game.rb]: :+1:

File: test/connect_four_test.rb [hunk 4]
```diff
@@ -1,7 +1,25 @@
...
+          return val
+        else
+          raise 'TestCase pulls more values than expected'
+        end
+      end
+    end
+  end
+
+  it "should play the game" do
+    game = ConnectFour.main
```
Comment (splattael) [inline test/connect_four_test.rb]: I like the idea of an integration test :+1: 

This test, however, produces output during `rake test`, so it's a bit noisy.

You could inject "input" and "output" into `ConnectFour.main`.

WDYT?

File: test/game/board_test.rb [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    8.times do
+      expected_output << "........\n"
+    end
+
+    assert_equal @test_board.to_s, expected_output
+  end
+
+  def test_solver
+    #diagonal
+    @test_board.fields = Matrix[%w[. . . . . . . .],
```
Comment (splattael) [inline test/game/board_test.rb]: I am not sure allowing users to _write_ `fields`.

We could hide that fact, that we use `Matrix` to implement `fields`.

WDYT?

File: lib/game/board.rb (hunk 1 above)
Comment (mark-a) [inline lib/game/board.rb]: yeah, i am aware of that.
 I thought for a second to use NMatrix from SciRuby, but didnt want to include this big dependency of an gcc compiled gem. 
Another idea was to reopen the basic class via inheritance like this:

```
class MutableMatrix < Matrix
 def []=(x, y, value)
    @rows[x][y] = value
  end
end
```
  


As im just using rows and cols from matrix i could also be better of with coming up with my own class and implementation.
What do you think should be the prefered way?

Comment (splattael) [inline lib/game/board.rb]: Actually, implementing `MutableMatrix` might work :)

I'd just make the method `[]=` `public`:

```ruby
class MutableMatrix < Matrix
  public :[]=
end
```

Remember that `Matrix` is used in tests to assign `fields`. The change above has to be addressed there as well, I suppose.

File: test/connect_four_test.rb (hunk 4 above)
Comment (mark-a) [inline test/connect_four_test.rb]: You are right, buffering console input and output internal to this test is the way to go.

File: test/game/board_test.rb (hunk 5 above)
Comment (mark-a) [inline test/game/board_test.rb]: Jupp i could implement a proxy "fill" function into the Board class.

**Pull request closed without merging**